### PR TITLE
check

### PR DESCRIPTION
**Pull request by Rotem19980**

### CODE REVIEW

File: functions.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
+import csv
+import datetime
+import exceptions_tests
+import os #this is needed for using directory paths and manipulating them
```
Comment (stavakler) [inline functions.py]: You don’t need to write this comment after import os:)

File: functions.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+import sys
+from pip._vendor.distlib.compat import raw_input
+
+class Employee(object):
+
+    def __init__(self, employee_id, name, phone, age):
```
Comment (stavakler) [inline functions.py]: Pls write document on the parameters that the function gets

File: functions.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+        self.employee_id = employee_id
+        self.name = name
```
Comment (stavakler) [inline functions.py]: Is it full name? Or first name?

File: functions.py [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+        self.phone = phone
+        self.age = age
+
+class Employees_list(object):
+    def __init__(self, list_of_employees):
```
Comment (stavakler) [inline functions.py:38]: Pls write document on the parameters that function gets

File: functions.py [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+        self.list_of_employees = list_of_employees
+
+    def add_employee(self, new_employee):
+        """
+        This function gets the credentials of a new employee from user and adds it to the employees file.
+        """
+        with open(self.list_of_employees, mode='a') as csvFile:
```
Comment (stavakler) [inline functions.py]: If the employee is already existing?

File: functions.py [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+        self.phone = phone
+        self.age = age
+
+class Employees_list(object):
+    def __init__(self, list_of_employees):
+        self.list_of_employees = list_of_employees
+
+    def add_employee(self, new_employee):
+        """
+        This function gets the credentials of a new employee from user and adds it to the employees file.
```
Comment (stavakler) [inline functions.py]: Pls add to this document the parameter that the function get.. (what new_employee is? and what type is it?)

File: functions.py [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+        """
+        with open(self.list_of_employees, mode='a') as csvFile:
+            fileout = csv.writer(csvFile, delimiter=',', quoting=csv.QUOTE_ALL)
+            fileout.writerow([new_employee.employee_id, new_employee.name, new_employee.phone, new_employee.age])
+
+    def add_employees_from_file(self):
+        """
+        This function gets from the user a file path that contains the data of the employees and adds it to the employees file
+        only if all the data of all employees is supplied.
+        new_employees = a string.
```
Comment (stavakler) [inline functions.py]: I don’t understand this line, there is no new_employee as parameter...

File: functions.py [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+        """
+        new_employees_file_path = raw_input("Please add the full file path of the employees you wish to add: ")
+        try:
+            with open(new_employees_file_path) as infile:
+                f = open(new_employees_file_path, 'r+')
+                data = f.read()
+                f.close()
+                f = open("employees.csv", "a")
```
Comment (stavakler) [inline functions.py]: Separate the read file and adding file to 2 different functions

File: functions.py [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+                    return row
+
+    # to do
+
+def monthly_attendance_report():
+    """
+    The function prints the attendance data of all employees from the last month.
+    """
+    this_month = datetime.datetime.now().month
+    with open('attendance_log.csv', 'r') as csvfile:
```
Comment (stavakler) [inline functions.py]: Pls put the path in variable

File: functions.py [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+    def mark_attendance(self):
+        """
+        The function gets an employee's id as the input and saves the date and time of him in the attendance log file.
+        employee_id = a 9 numbers integer.
+        """
+        self.employee_id = exceptions_tests.employee_id_test()
+        csv_data = list()
+        csv_data.append(self.employee_id)
+        csv_data.append(datetime.datetime.now())
+        with open('employees.csv') as File:
```
Comment (stavakler) [inline functions.py]: Pls put path in variable

File: functions.py [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+
+    # Done
+
+    def attendance_report_of_employee(self):
+        """
+        The function gets an employee's id as the input and prints all the entries of his attendance.
+        id_input = a 9 numbers integer.
+        """
+        self.employee_id = exceptions_tests.employee_id_test()
+        with open('attendance_log.csv', 'r') as csvfile:
```
Comment (stavakler) [inline functions.py]: Pls put path in variable☺️

File: functions.py [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+        csv_data.append(self.employee_id)
+        csv_data.append(datetime.datetime.now())
+        with open('employees.csv') as File:
+            reader = csv.reader(File, delimiter=',', quotechar=',',
+                                quoting=csv.QUOTE_MINIMAL)
+            for row in reader:
+                name = row[1]
+                if self.employee_id == row[0]:
+                    csv_data.append(name)
+        with open("attendance_log.csv", "a") as csvFile:
```
Comment (stavakler) [inline functions.py]: Pls put path in variable

File: functions.py [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+
+    #how to check if all data is supplied in the file
+
+    def delete_employee(self):
+        """
+        This function gets from the user a name of an employee he wishes to remove, writes all the rest of the employees to a temporary
+        file and then deletes the old employees file. Afterwards, it saves it as the main employees file (employees).
+        employees_name = a string.
+        """
+        self.name = exceptions_tests.name()
```
Comment (stavakler) [inline functions.py]: Something in this line don’t look right, Did you meant to give the function ‘exceptions_test’ the name? And what do you except to get as return?

File: functions.py [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+        with open("employees_edit.csv", "w") as my_empty_csv:
```
Comment (stavakler) [inline functions.py]: What is it?

File: functions.py [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+            with open(employees_to_delete_file_path, "r") as f:
+                data_to_erase = f.readlines()  # read data line by line
+            with open("employees.csv", "r") as f:
+                data_to_keep = f.readlines()  # read data line by line
+
+                # open file in write mode
+            with open("employees.csv", "w") as f:
+                for line in data_to_keep:
+                    if line in data_to_erase:
+                        pass
```
Comment (stavakler) [inline functions.py]: Why there is pass in here?

**Review by stavakler — COMMENTED** [review]
I will continue tomorrow:)

File: functions.py (hunk 3 above)
Comment (Rotem19980) [reply functions.py]: It's the full name of the employee - I need to add it to the name function in exceptions test file :)

File: functions.py (hunk 15 above)
Comment (Rotem19980) [reply functions.py]: data_to_keep is the data in the employee's file and. data_to_erase is the data we want to delete. I rewrite the employee's file by comparing between the two. for each line in the employee's file we check if it exists in the data_to_erase file - if so we pass it and not rewrite it to the employees file.

File: functions.py (hunk 5 above)
Comment (Rotem19980) [reply functions.py]: I tried to add a check to the function but it runs properly without doing the check, I'd be happy if you could take a look and see what's wrong in my command:

```    def add_employee(self, new_employee):
        new_employee_data = [new_employee.employee_id, new_employee.name, new_employee.phone, new_employee.age]
        with open(self.list_of_employees) as file1:
            existing_lines = csv.reader(file1)
            for row in existing_lines:
                if new_employee_data not in row:
                    new_employee_data.append(row)
                else:
                    print("Sorry, the employee is already exist.")
```

File: functions.py (hunk 9 above)
Comment (Rotem19980) [reply functions.py]: done

File: functions.py (hunk 10 above)
Comment (Rotem19980) [reply functions.py]: done

File: functions.py (hunk 11 above)
Comment (Rotem19980) [reply functions.py]: done :)

File: functions.py (hunk 12 above)
Comment (Rotem19980) [reply functions.py]: done

File: functions.py (hunk 3 above)
Comment (stavakler) [reply functions.py]: change from 'name' to 'full_name'

File: functions.py [hunk 16]
```diff
@@ -1,35 +1,50 @@
...
-        This function gets the credentials of a new employee from user and adds it to the employees file.
-        """
-        with open(self.list_of_employees, mode='a') as csvFile:
-            fileout = csv.writer(csvFile, delimiter=',', quoting=csv.QUOTE_ALL)
-            fileout.writerow([new_employee.employee_id, new_employee.name, new_employee.phone, new_employee.age])
+        new_employee_data = [new_employee.employee_id, new_employee.name, new_employee.phone, new_employee.age]
+        with open(self.list_of_employees, 'r+') as file1:
+            existing_lines = csv.reader(file1)
+            for row in existing_lines:
+                if new_employee_data in row:
```
Comment (stavakler) [inline functions.py]: maybe check only for ID and not all the details -> 
if  new_employee_data[0] in row: 
---------------------------------------------
if it still doesn't work try to print the row and see what inside it....

File: functions.py [hunk 17]
```diff
@@ -1,35 +1,50 @@
...
 
     def add_employee(self, new_employee):
-        """
-        This function gets the credentials of a new employee from user and adds it to the employees file.
-        """
-        with open(self.list_of_employees, mode='a') as csvFile:
-            fileout = csv.writer(csvFile, delimiter=',', quoting=csv.QUOTE_ALL)
-            fileout.writerow([new_employee.employee_id, new_employee.name, new_employee.phone, new_employee.age])
+        new_employee_data = [new_employee.employee_id, new_employee.name, new_employee.phone, new_employee.age]
+        with open(self.list_of_employees, 'r+') as file1:
```
Comment (stavakler) [inline functions.py]: pls change the name 'file1'

File: functions.py [hunk 18]
```diff
@@ -1,35 +1,50 @@
...
+            existing_lines = csv.reader(file1)
+            for row in existing_lines:
+                if new_employee_data in row:
+                    print("Sorry, the employee is already exist.")
```
Comment (stavakler) [inline functions.py]: message should be in const.

File: functions.py [hunk 19]
```diff
@@ -1,35 +1,50 @@
...
+            file1.write(str(new_employee_data) + '\n')
 
     def add_employees_from_file(self):
         """
         This function gets from the user a file path that contains the data of the employees and adds it to the employees file
         only if all the data of all employees is supplied.
-        new_employees = a string.
+        new_employees_file_path = a string.
         """
         new_employees_file_path = raw_input("Please add the full file path of the employees you wish to add: ")
```
Comment (stavakler) [inline functions.py]: message should be in const.

File: functions.py [hunk 20]
```diff
@@ -97,19 +112,21 @@ def mark_attendance(self):
         """
         The function gets an employee's id as the input and saves the date and time of him in the attendance log file.
         employee_id = a 9 numbers integer.
+        monthly_report_url = a string.
         """
+        monthly_report_url = input("Please enter monthly attendance url: ")
         self.employee_id = exceptions_tests.employee_id_test()
         csv_data = list()
```
Comment (stavakler) [inline functions.py]: csv_data = [] 
more appropriate :)

File: functions.py [hunk 21]
```diff
@@ -0,0 +1,183 @@
...
+def late_employees_report():
+    """
+    The function prints the attendance data of all employees who were late (came after 09:30).
+    """
+    with open(r'C:\Users\rotem\PycharmProjects\final_project_python\attendance_log.csv', 'r') as csvfile:
+        content = csv.reader(csvfile, delimiter=',')
+        for row in content:
+            if row:
+                hour = str(row[1].split(' ')[1])
+                if hour > "09:30:00":
```
Comment (stavakler) [inline functions.py]: "09:30:00" should be in const

File: functions.py (hunk 3 above)
Comment (Rotem19980) [reply functions.py]: done :)

File: functions.py (hunk 20 above)
Comment (Rotem19980) [reply functions.py]: done

File: functions.py (hunk 17 above)
Comment (Rotem19980) [reply functions.py]: done

File: functions.py (hunk 5 above)
Comment (Rotem19980) [reply functions.py]: done :)

File: functions.py (hunk 16 above)
Comment (Rotem19980) [reply functions.py]: done

File: functions.py (hunk 21 above)
Comment (Rotem19980) [reply functions.py]: done

File: functions.py (hunk 19 above)
Comment (Rotem19980) [reply functions.py]: done

File: functions.py (hunk 18 above)
Comment (Rotem19980) [reply functions.py]: done

File: functions.py (hunk 13 above)
Comment (Rotem19980) [reply functions.py]: I don't expect to get from the function something, it meant to check if the name is a valid value

Comment (stavakler) [reply functions.py]: You do have a return in 'name' function, but It's OK. pls change the name of the function 'name' that not good name, at first I didn't understand what is it...

File: functions.py (hunk 15 above)
Comment (stavakler) [reply functions.py]: You can't use 'pass' in the code, we are using 'pass' while we are writing a code, for our convenience, but we don't keep it when we finish to write the code. -> 
if line not in data_to_erase:
    f.write(line)

Comment (Rotem19980) [reply functions.py]: okay, understood and fixed :)

File: functions.py (hunk 14 above)
Comment (Rotem19980) [reply functions.py]: what exactly do you mean? the function checks if the employee we wish to delete exists in the file, and if he does, it removes him.

File: functions.py (hunk 13 above)
Comment (Rotem19980) [reply functions.py]: I renamed it :)

File: exceptions_tests.py [hunk 22]
```diff
@@ -0,0 +1,87 @@
...
+def csv_data_is_supplied_test():
+   try:
+       csv_path_from_user = raw_input("Please insert the csv path: ")
+       test = pd.read_csv(csv_path_from_user)
+       x = test.isnull().any(axis=1)
+       missing_data_rows = x[x == True].index.to_numpy()
+       if missing_data_rows.size > 0:
+           print("There's some data missing, please check your csv.")
+   except FileNotFoundError:
+       print('File does not exist.')
```
Comment (stavakler) [inline exceptions_tests.py]: CONST

File: exceptions_tests.py [hunk 23]
```diff
@@ -0,0 +1,87 @@
+import pandas as pd
+from pip._vendor.distlib.compat import raw_input
+
+def csv_data_is_supplied_test():
+   try:
+       csv_path_from_user = raw_input("Please insert the csv path: ")
```
Comment (stavakler) [inline exceptions_tests.py]: CONST

File: exceptions_tests.py [hunk 24]
```diff
@@ -0,0 +1,87 @@
...
+       test = pd.read_csv(csv_path_from_user)
+       x = test.isnull().any(axis=1)
```
Comment (stavakler) [inline exceptions_tests.py]: ``      x = test.isnull().any(axis=1)
 missing_data_rows = x[x == True].index.to_numpy()
       if missing_data_rows.size > 0:
           print("There's some data missing, please check your csv.") 
" 
can be much shoter...

File: exceptions_tests.py [hunk 25]
```diff
@@ -0,0 +1,87 @@
+import pandas as pd
+from pip._vendor.distlib.compat import raw_input
+
+def csv_data_is_supplied_test():
+   try:
+       csv_path_from_user = raw_input("Please insert the csv path: ")
+       test = pd.read_csv(csv_path_from_user)
```
Comment (stavakler) [inline exceptions_tests.py]: read_csv and raw_input can't be part of this function. pls seperete them to diffrent functions...

File: exceptions_tests.py [hunk 26]
```diff
@@ -0,0 +1,87 @@
...
+       print('File does not exist.')
+
+def employee_id_test():
+    """
+    This function gets from the user the employee's id as a string and checks if it's valid.
+    employee_id = a string.
+    """
+    while True:
+        try:
+            employee_id = str(input("Please add the employee's id: "))
```
Comment (stavakler) [inline exceptions_tests.py]: you can't put input inside the function... + CONST

File: exceptions_tests.py [hunk 27]
```diff
@@ -0,0 +1,87 @@
...
+            if any(char.isdigit() for char in employee_id) == True:
```
Comment (stavakler) [inline exceptions_tests.py]: it's can be :
if any(char.isdigit() for char in employee_id):

_without the == True_

File: exceptions_tests.py [hunk 28]
```diff
@@ -0,0 +1,87 @@
...
+                if len(str(employee_id)) == 9:
+                    if int(employee_id) < 100000000:
+                        employee_id = str(employee_id.zfill(9))
+                    return str(employee_id)
+                else:
+                    return 'Error! Please insert a 9 numbers employee id.'
```
Comment (stavakler) [inline exceptions_tests.py]: CONST

File: exceptions_tests.py [hunk 29]
```diff
@@ -0,0 +1,87 @@
...
+            else:
+                return 'Error! Please insert a 9 numbers employee id.'
```
Comment (stavakler) [inline exceptions_tests.py]: CONST

File: exceptions_tests.py [hunk 30]
```diff
@@ -0,0 +1,87 @@
...
+        except ValueError:
+            return 'Error! Please insert a valid employee id.'
```
Comment (stavakler) [inline exceptions_tests.py]: CONST

File: exceptions_tests.py [hunk 31]
```diff
@@ -0,0 +1,87 @@
...
+
+def full_name_test():
+    """
+    This function gets from the user a name of an employee as a string and checks if it's valid.
+    name = a string.
+    """
+    while True:
```
Comment (stavakler) [inline exceptions_tests.py]: I usually  not use while True... pls put condition instead 
like
 while any(char.isdigit() for char in name):

File: exceptions_tests.py [hunk 32]
```diff
@@ -0,0 +1,87 @@
...
+        try:
+            name = str(input("Please add the employee's name: "))
```
Comment (stavakler) [inline exceptions_tests.py]: Const

File: exceptions_tests.py [hunk 33]
```diff
@@ -0,0 +1,87 @@
...
+            if any(char.isdigit() for char in name) == False:
+                return str(name)
+            else:
+                return 'Error! Please insert a valid employee name.'
```
Comment (stavakler) [inline exceptions_tests.py]: CONST

File: exceptions_tests.py [hunk 34]
```diff
@@ -0,0 +1,87 @@
...
+
+def full_name_test():
+    """
+    This function gets from the user a name of an employee as a string and checks if it's valid.
+    name = a string.
+    """
+    while True:
+        try:
+            name = str(input("Please add the employee's name: "))
+            if any(char.isdigit() for char in name) == False:
```
Comment (stavakler) [inline exceptions_tests.py]: instead use - 
if not(any(char.isdigit() for char in name)):

File: exceptions_tests.py [hunk 35]
```diff
@@ -0,0 +1,87 @@
...
+            return 'Error! Please insert a valid employee name.'
+
+# to check that we get a full name
+
+def phone_test():
+    """
+    This function gets from the user the employee's phone as a string and checks if it's valid.
+    phone = a string.
+    """
+    while True:
```
Comment (stavakler) [inline exceptions_tests.py]: נ"ל

File: exceptions_tests.py [hunk 36]
```diff
@@ -0,0 +1,87 @@
...
+                return 'Error! Please insert a 10 digits phone number.'
+        except ValueError:
+            return 'Error! Please insert a valid phone number.'
+
+def age_test():
+    """
+    This function gets from the user an age of an employee as a integer and checks if it's valid.
+    age = an integer.
+    """
+    while True:
```
Comment (stavakler) [inline exceptions_tests.py]: נ"ל

File: exceptions_tests.py [hunk 37]
```diff
@@ -0,0 +1,87 @@
...
+                if len(str(phone)) == 10:
+                    if int(phone) < 100000000:
+                        phone = str(phone.zfill(9))
+                    return str(phone)
+                else:
+                    return 'Error! Please insert a 10 digits phone number.'
+            else:
+                return 'Error! Please insert a 10 digits phone number.'
+        except ValueError:
+            return 'Error! Please insert a valid phone number.'
```
Comment (stavakler) [inline exceptions_tests.py]: CONST

File: exceptions_tests.py [hunk 38]
```diff
@@ -0,0 +1,87 @@
...
+    This function gets from the user an age of an employee as a integer and checks if it's valid.
+    age = an integer.
+    """
+    while True:
+        try:
+            age = int(input("Please add the employee's age: "))
+            if 18 < age < 100:
+                return str(age)
+            else:
+                return "Sorry, please insert a valid age."
```
Comment (stavakler) [inline exceptions_tests.py]: CONST

File: functions.py [hunk 39]
```diff
@@ -0,0 +1,191 @@
...
+        This function gets the credentials of a new employee and employees file path from from user and adds it to the file.
+        new_employee = new employee's credentials.
+        """
+        with open(self.list_of_employees, 'r+') as list_of_employees_file:
+            list_of_employees_data = csv.reader(list_of_employees_file)
+            for row in list_of_employees_data:
+                if new_employee.employee_id in row:
+                    print(EMPLOYEE_ALREADY_EXISTS_MSG)
+                    return
+            list_of_employees_file.write(str(new_employee.employee_id) + ', ' + new_employee.name +
```
Comment (stavakler) [inline functions.py]: instead of concate (+) use string foramtting 
:)

File: functions.py [hunk 40]
```diff
@@ -0,0 +1,191 @@
...
+        self.full_name = exceptions_tests.full_name_test()
+        with open("employees_edit.csv", "w") as my_empty_csv:
+            pass
+        with open('employees.csv', 'r') as inp, open('employees_edit.csv', 'w') as out:
+            writer = csv.writer(out)
+            for row in csv.reader(inp):
+                if row[1] != self.name:
+                    writer.writerow(row)
+                else:
+                    return "Sorry, that employee does not exist in this file."
```
Comment (stavakler) [inline functions.py]: const

**Review by stavakler — COMMENTED** [review]
hi 
pls don't put 'input' inside the functions.
You are one the highway!!! :)

File: exceptions_tests.py (hunk 22 above)
Comment (Rotem19980) [reply exceptions_tests.py]: done

File: exceptions_tests.py (hunk 23 above)
Comment (Rotem19980) [reply exceptions_tests.py]: done

File: exceptions_tests.py (hunk 28 above)
Comment (Rotem19980) [reply exceptions_tests.py]: done

File: exceptions_tests.py (hunk 29 above)
Comment (Rotem19980) [reply exceptions_tests.py]: done

File: exceptions_tests.py (hunk 30 above)
Comment (Rotem19980) [reply exceptions_tests.py]: done

File: functions.py (hunk 40 above)
Comment (Rotem19980) [reply functions.py]: done

File: exceptions_tests.py (hunk 38 above)
Comment (Rotem19980) [reply exceptions_tests.py]: done

File: exceptions_tests.py (hunk 37 above)
Comment (Rotem19980) [reply exceptions_tests.py]: done

File: exceptions_tests.py (hunk 34 above)
Comment (Rotem19980) [reply exceptions_tests.py]: done

File: exceptions_tests.py (hunk 27 above)
Comment (Rotem19980) [reply exceptions_tests.py]: done

File: functions.py (hunk 39 above)
Comment (Rotem19980) [reply functions.py]: done :)

File: exceptions_tests.py (hunk 26 above)
Comment (Rotem19980) [reply exceptions_tests.py]: done

File: exceptions_tests.py (hunk 31 above)
Comment (Rotem19980) [reply exceptions_tests.py]: done

File: exceptions_tests.py (hunk 33 above)
Comment (Rotem19980) [reply exceptions_tests.py]: done

File: exceptions_tests.py (hunk 32 above)
Comment (Rotem19980) [reply exceptions_tests.py]: done

File: exceptions_tests.py (hunk 36 above)
Comment (Rotem19980) [reply exceptions_tests.py]: done

File: exceptions_tests.py (hunk 35 above)
Comment (Rotem19980) [reply exceptions_tests.py]: done

File: exceptions_tests.py (hunk 25 above)
Comment (Rotem19980) [reply exceptions_tests.py]: done :)

File: exceptions_tests.py (hunk 24 above)
Comment (Rotem19980) [reply exceptions_tests.py]: done :)

File: functions.py (hunk 8 above)
Comment (Rotem19980) [reply functions.py]: done :)

File: functions.py (hunk 14 above)
Comment (Rotem19980) [reply functions.py]: done

File: exceptions_tests.py [hunk 41]
```diff
@@ -0,0 +1,102 @@
...
+EMPLOYEE_INVALID_ID_MSG = 'Error! Please insert a 9 numbers employee id.'
+EMPLOYEE_INVALID_AGE_MSG = "Sorry, please insert a valid age."
+EMPLOYEE_INVALID_PHONE_MSG = 'Error! Please insert a valid phone number.'
+EMPLOYEE_ID_INPUT_MSG = "Please add the employee's id: "
+EMPLOYEE_INVALID_NAME_MSG = 'Error! Please insert a valid employee name.'
+EMPLOYEE_NAME_INPUT_MSG = "Please add the employee's name: "
+EMPLOYEE_AGE_INPUT_MSG = "Please add the employee's age: "
+EMPLOYEE_PHONE_INPUT_MSG = "Please add the employee's phone: "
+
+def open_and_read_file(csv_file):
```
Comment (stavakler) [inline exceptions_tests.py]: pls write only ' def read_file(csv_file):'

File: functions.py [hunk 42]
```diff
@@ -0,0 +1,201 @@
...
+class EmployeesList(object):
+
+    def __init__(self, list_of_employees):
+        """
+        This function gets employees file's path from user.
+        list_of_employees = a string.
+        """
+        self.list_of_employees = list_of_employees
+
+    def open_and_read_file(self, list_of_employees):
```
Comment (stavakler) [inline functions.py]: def read_file(self, list_of_employees): 

better name

File: functions.py [hunk 43]
```diff
@@ -0,0 +1,201 @@
...
+        """
+        This function gets employees file's path from user and returns its content.
+        list_of_employees = a string.
+        """
+        employees_csv_data = pd.read_csv(list_of_employees)
+        return employees_csv_data
+
+    def open_and_write_employee_file(self, new_employee):
```
Comment (stavakler) [inline functions.py]: def write_employee_file(self, new_employee):'

better name

File: functions.py [hunk 44]
```diff
@@ -0,0 +1,201 @@
...
+from pip._vendor.distlib.compat import raw_input
+
+EMPLOYEE_DOESNT_EXIST_MSG = "Sorry, the employee doesn't exist."
+EMPLOYEE_ALREADY_EXISTS_MSG = "Sorry, the employee is already exist."
+FILE_DOES_NOT_EXIST_MSG = "File does not exist."
+LATE_HOUR = "09:30:00"
+ADD_EMPLOYEES_FROM_FILE_MSG = "Please add the full file path of the employees you wish to add: "
+EMPLOYEE_ATTENDANCE_DOESNT_EXIST_MSG = "Sorry, the employee has no data in the attendance report."
+EMPLOYEE_DOESNT_EXIST_IN_FILE_MSG = "Sorry, that employee does not exist in this file."
+EMPLOYEES_EDIT_FILE = r'C:\Users\rotem\PycharmProjects\final_project_python\employees_edit.csv'
```
Comment (stavakler) [inline functions.py]: EMPLOYEES_EDIT_PATH 

better name

File: functions.py [hunk 45]
```diff
@@ -0,0 +1,201 @@
...
+REMOVE_EMPLOYEES_FROM_FILE_MSG = "Please add the file path of the employees you wish to remove: "
+ATTENDANCE_LOG_FILE_PATH = r'C:\Users\rotem\PycharmProjects\final_project_python\attendance_log.csv'
```
Comment (stavakler) [inline functions.py]: ATTENDANCE_LOG_PATH

better name

File: functions.py [hunk 46]
```diff
@@ -0,0 +1,201 @@
+import csv
+import datetime
+import exceptions_tests
+import os
+import pandas
```
Comment (stavakler) [inline functions.py]: duplicates lines

File: functions.py [hunk 47]
```diff
@@ -0,0 +1,201 @@
...
+            print(FILE_DOES_NOT_EXIST_MSG)
+
+    def delete_employee(self, new_employee):
+        """
+        This function gets from the user a name of an employee he wishes to remove, writes all the rest of the
+        employees to a temporary file and then deletes the old employees file. Afterwards, it saves it as the main
+        employees file (employees). employees_name = a string.
+        """
+        employees_csv_data = self.open_and_read_file(self.list_of_employees)
+        with open(EMPLOYEES_EDIT_FILE, 'w') as updated_csv:
```
Comment (stavakler) [inline functions.py]: which file is that?

File: functions.py [hunk 48]
```diff
@@ -0,0 +1,201 @@
...
+        attendance_time = attendance_time.strftime('%Y/%m/%d %H:%M:%S')
+        attendance_time = str(attendance_time)
+        attendance_data.append(attendance_time)
+        with open(self.list_of_employees) as File:
+            reader = csv.reader(File, delimiter=',', quotechar=',',
+                                quoting=csv.QUOTE_MINIMAL)
+            for row in reader:
+                if row:
+                    if str(new_employee.employee_id) == str(row[0]):
+                        name_to_save = row[1]
```
Comment (stavakler) [inline functions.py]: """
 name_to_save = row[1]
 attendance_data.append(str(name_to_save))
"""

pls write in one line

**Review by stavakler — COMMENTED** [review]
excellent!!!!!!11 :))

File: functions.py (hunk 42 above)
Comment (Rotem19980) [reply functions.py]: done :)

File: exceptions_tests.py (hunk 41 above)
Comment (Rotem19980) [reply exceptions_tests.py]: done

File: functions.py (hunk 43 above)
Comment (Rotem19980) [reply functions.py]: done

File: functions.py (hunk 44 above)
Comment (Rotem19980) [reply functions.py]: done

File: functions.py (hunk 45 above)
Comment (Rotem19980) [reply functions.py]: done

File: functions.py (hunk 48 above)
Comment (Rotem19980) [reply functions.py]: done

File: functions.py (hunk 46 above)
Comment (Rotem19980) [reply functions.py]: done

File: functions.py (hunk 47 above)
Comment (Rotem19980) [reply functions.py]: temporary file for rewriting the updated data within

**Pull request merged**